Add home tasks backend basic structure.

aaronfc opened the pull request:
Fixes #5739 

### Changes proposed in this Pull Request
* [x] Add Tasks structure in the backend.
* [x] Implement sample "Site Setup" task.
* [x] Integrate provider with mapper and REST API.
* [ ] ~Add logic based on purposes to decide on tasks.~ Will be implemented on each separate task.
* [x] Allow for other plugins to modify returned tasks.
* [ ] ~All the individual tasks~ Will be implemented in separate tickets.

### Testing instructions
- Similar to previous tasks (#5739 #5741 #5627 ).
- Check that the "Setup Site" task is returned.

### Screenshot / Video
```
$ curl -s -X GET --user aaron:development --insecure 'https://test2.local/wp-json/sensei-internal/v1/home' | jq .tasks
{
  "items": {
    "setup-site": {
      "title": "Set up Course Site",
      "priority": 100,
      "url": "https://test2.local/wp-admin/admin.php?page=sensei_setup_wizard",
      "image": null,
      "done": true
    }
  }
}
```

aaronfc committed:
Add structure for Tasks.

aaronfc committed:
Integrate tasks provider with controller and mapper.

aaronfc committed:
Add id and priority fields to tasks. Add  filter.

- Added `get_id` method to the task interface.
- Added `get_priority` method to the task interface.
- Updated REST interface to return a dictionary indexed by task ID.
- Updated REST interface to return priority field.
- Added `sensei_home_tasks_items` filter.

fjorgemota commented on includes/rest-api/mappers/class-sensei-rest-api-home-controller-mapper.php:
```diff
@@ -112,4 +112,55 @@ public function map_promo_banner( Sensei_Home_Promo_Banner $banner ): array {
+	 * @return array
+	 */
+	public function map_tasks( Sensei_Home_Tasks $tasks ): array {
+		$mapped = [];
+		foreach ( $tasks->get_items() as $task ) {
+			$mapped[ $task->get_id() ] = $this->map_task( $task );
+		}
```
WDYT about maybe returning an array/list ordered by `priority` here, instead of mapped by ID?

aaronfc replied:
I thought about that, but that way it would be difficult for people using the filter to "remove" one, for example.
That's the main idea on having the ID, so we can uniquely identify a specific task.

fjorgemota replied:
Hey there, hmmm, good point. I think we would need to remove the task only on the backend, using the filter, right?

In this case, maybe the improvements [here](https://github.com/Automattic/sensei/pull/5815#discussion_r985902964) could help to remove the need for mapping by ID first?

fjorgemota commented on includes/rest-api/mappers/class-sensei-rest-api-home-controller-mapper.php:
```diff
@@ -112,4 +112,55 @@ public function map_promo_banner( Sensei_Home_Promo_Banner $banner ): array {
+		}
+
+		return [
+			/**
+			 * Filter to alter the list of tasks that will be sent to the frontend.
+			 * New tasks can be added and existing tasks modified or removed.
+			 * Each item must have:
+			 * - title: the title of the task.
+			 * - priority: number used in frontend to sort tasks.
+			 * - url: optional string with a url users will be taken when clicking on the task.
+			 * - image: optional string with a path/url to a featured image.
+			 * - done: boolean to mark task as done or not.
+			 *
+			 * @hook sensei_home_tasks_items
+			 * @since $$next-version$$
+			 *
+			 * @param array $tasks An array indexed by task ID.
+			 *
+			 * @return {bool}
+			 */
+			'items' => apply_filters( 'sensei_home_tasks_items', $mapped ),
```
WDYT about maybe applying this filter directly on `$tasks`, and adding a new method like `add_task` on `Sensei_Home_Tasks`, so we can add tasks properly and have them properly formatted here automatically for us?

aaronfc replied:
I am glad you raised this, because I don't like the result and maybe I did some wrong assumption:
I did this in this way because I didn't want to expose `Sensei_Home_Tasks` (nor `Sensei_Home_Task`) since I am assuming that other plugin developers might not have access to those classes.

So that's why I am applying the filter AFTER mapping to a plain array structure. What do you think on this? Is it safe to assume that they will have access to any of those classes?

fjorgemota replied:
Hey there. Very good point indeed!

WDYT about maybe having a class that implements `Sensei_Home_Task`, but returns the data from an array instead, similar to the structure that this filter needs to provide?

I'm asking because at least, in this case, we would be able to provide an `add_task` method on `Sensei_Home_Tasks`, and then we would have some validation on the array format, which is good when working with these filters.

In the same way, we would be able to provide a `remove_task` method, to remove a given task by ID, too, which would be nice.

WDYT?

aaronfc replied:
I am sorry Fernando, but I am not sure I am following you correctly. I will open an slack with you so we can talk about this today :)

fjorgemota reviewed (commented):
This is looking good! I added a few comments with suggestions. :)

fjorgemota commented on includes/admin/home/tasks/class-sensei-home-task.php:
```diff
@@ -0,0 +1,57 @@
+ * @package sensei-lms
+ * @since $$next-version$$
+ */
+
```
```suggestion

if ( ! defined( 'ABSPATH' ) ) {
	exit; // Exit if accessed directly.
}

```

fjorgemota commented on includes/admin/home/tasks/class-sensei-home-tasks-provider.php:
```diff
@@ -0,0 +1,35 @@
+ * @package sensei-lms
+ * @since   $$next-version$$
+ */
+
```
```suggestion

if ( ! defined( 'ABSPATH' ) ) {
	exit; // Exit if accessed directly.
}

```

fjorgemota commented on includes/admin/home/tasks/class-sensei-home-tasks.php:
```diff
@@ -0,0 +1,40 @@
+ * @package sensei-lms
+ * @since $$next-version$$
+ */
+
```
```suggestion

if ( ! defined( 'ABSPATH' ) ) {
	exit; // Exit if accessed directly.
}

```

fjorgemota commented on includes/admin/home/tasks/task/class-sensei-home-task-setup-site.php:
```diff
@@ -0,0 +1,70 @@
+ * @package sensei-lms
+ * @since $$next-version$$
+ */
+
```
```suggestion

if ( ! defined( 'ABSPATH' ) ) {
	exit; // Exit if accessed directly.
}

```

fjorgemota commented on tests/unit-tests/admin/home/tasks/test-class-sensei-home-tasks-provider.php:
```diff
@@ -0,0 +1,43 @@
+ *
+ * @package sensei
+ */
+
```
```suggestion

if ( ! defined( 'ABSPATH' ) ) {
	exit; // Exit if accessed directly.
}
```

fjorgemota reviewed (changes requested):
I just noticed that this PR doesn't have the traditional  `ABSPATH` checks in some new files, so I suggested adding them. :)

More information: https://stackoverflow.com/questions/43212340/what-is-meant-by-if-defined-abspath

fjorgemota commented on includes/admin/home/tasks/class-sensei-home-tasks.php:
```diff
@@ -0,0 +1,40 @@
+	}
+
+	/**
+	 * The the actual tasks.
```
```suggestion
	 * The actual tasks.
```

Just a minor typo. :)

aaronfc committed:
Fix typo in comment.

Co-authored-by: Fernando Jorge Mota <fjorgemota@users.noreply.github.com>

aaronfc committed:
Apply suggestions from code review

Add ABSPATH check to new files.

Co-authored-by: Fernando Jorge Mota <fjorgemota@users.noreply.github.com>

aaronfc committed:
Update hook name and documentation.

- Renamed filter from `sensei_home_tasks_items` to `sensei_home_tasks`.
- Updated documentation to follow standard.
- Added ID to the task structure.

aaronfc committed:
Fix tests.

aaronfc committed:
Fix test message.

fjorgemota reviewed (approved):
Looks good and works well! :)
